<a href="https://colab.research.google.com/github/mSounak/Heart-Disease-UCI/blob/main/Heart_Diseases_Using_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Heart Diseases Classifier using K-Nearest Neighbors

About the dataset:<br>
This dataset has 13 features and 303 patient. It has a target value column which has values 1 which means has heart diseases and 0 means doesn't has heart diseases.

The features have age and
```
Sex:

0: Female
1: Male

Chest Pain Type:

0: Typical Angina
1: Atypical Angina
2: Non-Anginal Pain
3: Asymptomatic

Resting Blood Pressure: Person's resting blood pressure.
Cholesterol: Serum Cholesterol in mg/dl
Fasting Blood Sugar:

0:Less Than 120mg/ml
1: Greater Than 120mg/ml

Resting Electrocardiographic Measurement:

0: Normal
1: ST-T Wave Abnormality
2: Left Ventricular Hypertrophy

Max Heart Rate Achieved: Maximum Heart Rate Achieved
Exercise Induced Angina:

1: Yes
0: No

ST Depression: ST depression induced by exercise relative to rest.
Slope: Slope of the peak exercise ST segment:

0: Upsloping
1: Flat
2: Downsloping

Thalassemia: A blood disorder called 'Thalassemia':

0: Normal
1: Fixed Defect
2: Reversable Defect

Number of Major Vessels: Number of major vessels colored by fluoroscopy.
```

### Task 1: Importing all the necessary modules.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression

### Task 2: Importing Data

In [2]:
url = 'https://raw.githubusercontent.com/mSounak/Heart-Disease-UCI/main/heart.csv?token=AHH7EOP4WA7GSF7D3CB2CPDADDRSY'
data = pd.read_csv(url)
data.head()

age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]

### Task 3: Missing Data

In [3]:
data.dtypes

age           int64
sex           int64
cp            int64
trestbps      int64
chol          int64
fbs           int64
restecg       int64
thalach       int64
exang         int64
oldpeak     float64
slope         int64
ca            int64
thal          int64
target        int64
dtype: object

In [4]:
data.isna().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [5]:
data.shape

(303, 14)

In [6]:
type(data)

pandas.core.frame.DataFrame

### Task 4: Split the data

In [7]:
#store the target in 'y'

y = data["target"]
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

In [8]:
#store the features in 'X'

X = data.drop("target", axis = 1)
X.head()

age  sex  cp  trestbps  chol  fbs  ...  thalach  exang  oldpeak  slope  ca  thal
0   63    1   3       145   233    1  ...      150      0      2.3      0   0     1
1   37    1   2       130   250    0  ...      187      0      3.5      0   0     2
2   41    0   1       130   204    0  ...      172      0      1.4      2   0     2
3   56    1   1       120   236    0  ...      178      0      0.8      2   0     2
4   57    0   0       120   354    0  ...      163      1      0.6      2   0     2

[5 rows x 13 columns]

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 21, stratify = y)

In [10]:
X.describe()

age         sex          cp  ...       slope          ca        thal
count  303.000000  303.000000  303.000000  ...  303.000000  303.000000  303.000000
mean    54.366337    0.683168    0.966997  ...    1.399340    0.729373    2.313531
std      9.082101    0.466011    1.032052  ...    0.616226    1.022606    0.612277
min     29.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000
25%     47.500000    0.000000    0.000000  ...    1.000000    0.000000    2.000000
50%     55.000000    1.000000    1.000000  ...    1.000000    0.000000    2.000000
75%     61.000000    1.000000    2.000000  ...    2.000000    1.000000    3.000000
max     77.000000    1.000000    3.000000  ...    2.000000    4.000000    3.000000

[8 rows x 13 columns]

### Task 5: Normalization

In [11]:
scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

### Task 6: Train the model using K-NN 

In [12]:
model = KNeighborsClassifier(n_neighbors=8)

model.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=8, p=2,
                     weights='uniform')

### Task 7: Test the model

In [13]:
acc_KNN = model.score(X_test, y_test)*100

print("K-Nearest Neighbor Test Accuracy {:.2f}%".format(acc_KNN))

K-Nearest Neighbor Test Accuracy 88.52%


### Train the model using Logistic Regression

In [14]:
log_regr = LogisticRegression(max_iter = 1500)
log_regr.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=1500,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [15]:
acc_LR = log_regr.score(X_test, y_test)*100

print("Logistic Regression Test score {:.2f}%".format(acc_LR))

Logistic Regression Test score 83.61%


***We can clearly see KNN is giving us better accuracy with 88.52%***

### Task 8: Predict some data

In [19]:
predict_df = pd.DataFrame({
    "age" : [18],
    "sex" : [1],
    "cp" : [3],
    "trestbps" : [160],
    "chol" : [280],
    "fbs" : [1],
    "restecg" : [1],
     "thalach" : [182],
     "exang" : [1],
     "oldpeak" : [1.3],
     "slope" : [2],
     "ca" : [0],
     "thal" : [3],
})

In [20]:
predict_df

age  sex  cp  trestbps  chol  fbs  ...  thalach  exang  oldpeak  slope  ca  thal
0   18    1   3       160   280    1  ...      182      1      1.3      2   0     3

[1 rows x 13 columns]

In [21]:
predict_df = scaler.transform(predict_df)

model.predict(predict_df)

array([1])

**We can see that our model is predicting data perfectly.**